sdk/core/azure-core-v2/src/main/java/com/azure/v2/core/implementation/AccessTokenCache.java

```diff
@@ -71,69 +71,64 @@ public AccessToken getToken(TokenRequestContext tokenRequestContext, boolean che
     }
 
     private AccessToken retrieveToken(TokenRequestContext tokenRequestContext, boolean forceFetchToken) {
+        validateTokenRequestContext(tokenRequestContext);
+        return fetchToken(tokenRequestContext, forceFetchToken);
+    }
+
+    private void validateTokenRequestContext(TokenRequestContext tokenRequestContext) {
         if (tokenRequestContext == null) {
             throw LOGGER
                 .logThrowableAsError(new IllegalArgumentException("The token request context input cannot be null."));
         }
+    }
+
+    private AccessToken fetchToken(TokenRequestContext tokenRequestContext, boolean forceFetchToken) {
+        OffsetDateTime now = OffsetDateTime.now();
         AccessTokenCacheInfo cache = this.cacheInfo;
         AccessToken cachedToken = cache.getCachedAccessToken();
 
-        OffsetDateTime now = OffsetDateTime.now();
-        boolean tokenRefresh;
-        AccessToken fallback;
+        boolean needsRefresh = determineRefreshRequirement(tokenRequestContext, forceFetchToken, cachedToken, now);
 
-        // Check if the incoming token request context is different from the cached one. A different
-        // token request context, requires to fetch a new token as the cached one won't work for the
-        // passed in token request context.
-        boolean forceRefresh
-            = (forceFetchToken && checkIfForceRefreshRequired(tokenRequestContext)) || this.tokenRequestContext == null;
+        if (!needsRefresh && cachedToken != null) {
```

Comment:
Can remove the null check on `cachedToken` as `determineIfRefreshRequirement` also checks it.